engine-standalone-tracing/src/types/call_tracer.rs

```diff
@@ -0,0 +1,288 @@
+
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub struct CallFrame {
+    call_type: CallType,
+    from: Address,
+    to: Option<Address>,
+    value: U256,
+    gas: u64,
+    gas_used: u64,
+    input: Vec<u8>,
+    output: Vec<u8>,
+    error: Option<String>,
+    calls: Vec<CallFrame>,
+}
+
+#[derive(Debug, Default, Clone, PartialEq, Eq)]
+pub struct CallTracer {
+    call_stack: Vec<CallFrame>,
+}
+
+impl CallTracer {
+    fn end(&mut self, output: Vec<u8>, error: Option<&evm::ExitReason>) {
+        let frame = self.call_stack.first_mut().unwrap();
+        match error {
+            None => frame.output = output,
+            Some(error) => {
+                let error_message = format!("{:?}", error);
+                if error_message.to_lowercase().contains("revert") {
+                    frame.output = output;
+                }
+                frame.error = Some(error_message);
+            }
+        }
+    }
+
+    fn enter(
+        &mut self,
+        call_type: CallType,
+        from: Address,
+        to: Address,
+        input: Vec<u8>,
+        gas: u64,
+        value: U256,
+    ) {
+        let frame = CallFrame {
+            call_type,
+            from,
+            to: Some(to),
+            value,
+            gas,
+            gas_used: 0,
+            input,
+            output: Vec::new(),
+            error: None,
+            calls: Vec::new(),
+        };
+        self.call_stack.push(frame);
+    }
+
+    fn exit(&mut self, output: Vec<u8>, error: Option<&evm::ExitReason>) {
+        if self.call_stack.len() <= 1 {
+            return self.end(output, error);
+        }
+
+        let mut frame = self.call_stack.pop().unwrap();
+        match error {
+            None => frame.output = output,
+            Some(error) => {
+                frame.error = Some(format!("{:?}", error));
+                match frame.call_type {
+                    CallType::Create | CallType::Create2 => frame.to = None,
+                    _ => (),
+                }
+            }
+        }
+
+        self.call_stack.last_mut().unwrap().calls.push(frame);
+    }
+
+    fn update_gas_from_snapshot(&mut self, snapshot: Option<evm_gasometer::Snapshot>) {
+        if let Some(snapshot) = snapshot {
+            if let Some(frame) = self.call_stack.last_mut() {
+                frame.gas = snapshot.gas_limit;
+                frame.gas_used = snapshot.used_gas + snapshot.memory_gas;
+            }
+        }
+    }
+
+    #[cfg(feature = "serde")]
+    pub fn serializable(mut self) -> Option<SerializableCallFrame> {
+        if self.call_stack.len() != 1 {
+            // If there is more than one element in `call_stack` then it must mean the trace did not complete
+            // because there is only 1 top-level call. Note: additional frames are added as new scopes are entered,
+            // but then the frames are coalesced as those scopes are existed.
+            return None;
+        }
```

Comment:
I take it that this is the essentially [this](https://github.com/ethereum/go-ethereum/blob/ad15050c7fbedd0f05a49e81400de18c2cc2c284/eth/tracers/native/call.go#L105-L108)?

Reply:
No, it's [this](https://github.com/ethereum/go-ethereum/blob/ad15050c7fbedd0f05a49e81400de18c2cc2c284/eth/tracers/native/call.go#L152-L154).